onnxruntime/core/optimizer/embed_layer_norm_fusion.cc

```diff
@@ -0,0 +1,183 @@
+    if (p_attention == nullptr) {
+      continue;
+    }
+    Node& attention_node = *graph.GetNode(p_attention->Index());
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(attention_node, "Attention", {1}, kMSDomain) ||
+        !graph_utils::IsSupportedProvider(attention_node, GetCompatibleExecutionProviders()) ||
+        !IsSupportedDataType(attention_node)) {
+      continue;
+    }
+    // Find ReduceSum --> Attention
+    std::vector<const Node::EdgeEnd*> edges;
+    if (!graph_utils::FindPath(attention_node, true, {{0, 3, "ReduceSum", {1}, kOnnxDomain}}, edges, logger)) {
+      continue;
+    }
+    Node& reduce_sum_node = *graph.GetNode(edges[0]->GetNode().Index());
+    if (reduce_sum_node.GetOutputEdgesCount() != 1) {
+      continue;
+    }
+
+    // Traceback the SkipLayerNormalization node to find Gather --> SkipLayerNormalization
+    std::vector<graph_utils::EdgeEndToMatch> segment_embedding_path{
+        {0, 0, "Gather", {1}, kOnnxDomain}};
+    edges.clear();
+    if (!graph_utils::FindPath(skip_ln_node, true, segment_embedding_path, edges, logger)) {
+      continue;
+    }
+    Node& segment_gather_node = *graph.GetNode(edges[0]->GetNode().Index());
+    if (segment_gather_node.GetOutputEdgesCount() != 1) {
+      continue;
+    }
+
+    // Traceback the SkipLayerNormalization node to find Gather --> Add --> SkipLayerNormalization
+    std::vector<graph_utils::EdgeEndToMatch> word_embedding_path{
+        {0, 1, "Add", {7}, kOnnxDomain},
+        {0, 0, "Gather", {1}, kOnnxDomain}};
+    edges.clear();
+    if (!graph_utils::FindPath(skip_ln_node, true, word_embedding_path, edges, logger)) {
+      continue;
+    }
+    Node& add_node = *graph.GetNode(edges[0]->GetNode().Index());
+    Node& word_gather_node = *graph.GetNode(edges[1]->GetNode().Index());
+    if (add_node.GetOutputEdgesCount() != 1 || word_gather_node.GetOutputEdgesCount() != 1) {
+      continue;
+    }
+
+    // Traceback the Add node to find Shape --> Expand --> Gather --> Add
+    std::vector<graph_utils::EdgeEndToMatch> position_embedding_path{
+        {0, 1, "Gather", {1}, kOnnxDomain}};
+        //{0, 1, "Expand", {8}, kOnnxDomain},
+        //{0, 1, "Shape", {1}, kOnnxDomain}};
+    edges.clear();
+    if (!graph_utils::FindPath(add_node, true, position_embedding_path, edges, logger)) {
+      continue;
+    }
+    Node& position_gather_node = *graph.GetNode(edges[0]->GetNode().Index());
+    //Node& expand_node = *graph.GetNode(edges[1]->GetNode().Index());
+    //Node& shape_node = *graph.GetNode(edges[2]->GetNode().Index());
+    if (position_gather_node.GetOutputEdgesCount() != 1) {
+      continue;
+    }
+
+    // Get input "input_ids" from node.
+    NodeArg* input_ids = nullptr;
+    if (!graph_utils::NodeArgIsConstant(graph, *(word_gather_node.MutableInputDefs()[1])) &&
+        !graph_utils::IsGraphInput(graph, word_gather_node.MutableInputDefs()[1])) {
```

Comment:
Shall we add check of first input for 3 Gather nodes? For example, the input has the initializer of 2D tensor.